kernel/read.c

```diff
@@ -191,7 +261,8 @@ NTSTATUS DeviceRead(IN PDEVICE_EXTENSION extension,
 	mdl = IoAllocateMdl(toxic_buffer, current_read_window,  FALSE, TRUE, NULL); // <= toxic buffer address increases each time in the loop.
 	if (!mdl)
 	{
-		return STATUS_INSUFFICIENT_RESOURCES;
+		status = STATUS_INSUFFICIENT_RESOURCES;
+		goto end;

```

Comment:
if we are going to change the style of the function to goto end, i think we need to be consistent about moving all the other cleanup code to the tail no?

Reply:
Not necessarily, because different cleanup needs to be done depending on what failed at which state.
We can only put the mutex release at the tail.